### DISCUSSION TITLE
smartObjects[].resize not working (PS 24.1.0.166)

### DISCUSSION THREAD

**Discussion by spaceda1** [joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script#17] [Q&A]
Hi
I'm trying to make it working with PS 2023 (24.1.0.166)
none of resize values are working: 'fill', 'fit', 'xFill', 'yFill'
image always is mocked with proper height never fill or fit.
any idea how to fix it?
thank you
Zbig

**Comment by joonaspaakko** [comment]
What happens if you try the [standalone example](https://github.com/joonaspaakko/Batch-Mockup-Smart-Object-Replacement.js-Photoshop-script-standalone-example) just like that without changing anything? I tested it in CC 2022 + CC 2023 and both did fit and fill just fine. 

If you get the same result, I'd suggest you try replacing its input files with your own from the test where they didn't get properly resized. 

What I'm suspecting is that it could have something to do with the resolution of the input files (or the mockup). Another test could be to try and match their resolution to the mockup. 

If it seems hopeless, I can take a look at it, if you send me your whole project folder (my email is in my profile). At the very least include the mockup psd, one input file, and the settings script. If you don't want to share the the mockup or the input files, you could resize them to a small enough size that I can see what's wrong but can't really use them. You could also watermark them if you like. 

If you do modify either image file in any way, try the script on it yourself before sending them to me, just to make sure the issue can still be replicated in your end.

**Reply by spaceda1** [reply]
Hi
indeed standalone version worked. I used high resolution public domain images in any aspect ratio available. Cant tell what the difference between scripts is but standalone version worked.

So while it looks I can use resize feature now more difficult question how to align frame (fit) to the actual size / aspect ratio of image?
I can imagine there is a layer/s with frame we can manipulate after we get image file params. DO you recognize any script capable of such operations?

thank you for  Batch-Mockup - this is great tool!
regs
Zbig

**Reply by joonaspaakko** [reply]
Hmm... If the resize option works properly, the input images will be resized to fit or fill the target area (smart object bounds) while keeping the original aspect ratio. If it doesn't do that, there's likely a user error or a bug somewhere. 

If you're trying to use the script with a `Frame` in PS, that could be where the issue is coming from. Are you using someone else's mockup file or trying to make one yourself? The script works with `Smart Objects` only, which is what Photoshop mockup templates tend to use, just like the standalone example or any of the other examples [here](https://github.com/joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script/tree/master/Examples).

**Reply by spaceda1** [reply]
I have premium mockup psd files with smart objects inside. 
I'm not using any other scripts.
see example what I have at the moment  (1) and what I would like to get (2).
![2](https://user-images.githubusercontent.com/124141579/216493254-c9f0b9c4-480e-4561-94c0-c3be77c3af01.jpg)
![1](https://user-images.githubusercontent.com/124141579/216493256-0a40714f-0a20-4a0e-b3ca-7dc5895483e3.jpg)

**Reply by joonaspaakko** [reply]
Alright, and you're saying the `resize: "fit"` option didn't work? 

If that's the case I think I might have to take a look at your input file, the mockup file, and the settings script to figure it out...

**Reply by spaceda1** [reply]
Fit and fill both work. I'm just looking for a way to adjust frame to image instead image to the frame...
First image is manually modified by me just to show you what I would like to achieve. to add a frame to image instead of fill or fit image to the frame.

it does not need t to be this frame - can be black thin frame - just to find a script which is capable of playing with layer/mask/ smart object.
any idea how to make it?

**Reply by joonaspaakko** [reply]
Oh! You know I think the images you uploaded previously were flipped around... Or at least the description above says the second image is your desired outcome. I was so focused on the potential issue with the script that I didn't even consider you wanted something completely different and I failed to realize the frame on the first image was shorter. 

I see what you mean. Disregarding the mockup for now, I happen to have an [image framing `Action`](https://www.dropbox.com/sh/f5c8int3sqxy81p/AABuFX5iRc1Cvj-A7usjjxPUa?dl=0) I made for someone years ago. I modified the action just now to make it so the first `Open` and the `Place` actions need user action, so it's easier to test. I also disabled the last step, which is to export the image. In the first dialog you should pick `Frame1.psd` and on the second dialog you should pick `Frame Cutter.psd`. This `Action` was meant to be used in Photoshop's Batch process (`File > Automate > Batch...`), in which case you would have to disable user interactions for Open and Place, as well as change the file paths. Also the `Save > Export` needs a file path change. This would be way more flexible as a script, but script would've also been more work. 

I had a gif lying around that kind of explains how the action functions:
- It copy & pastes the frame image to all sides except left side
- It then uses the "Frame Cutter" triangle to cut miter joints to top and the bottom 
- As a final step it duplicates and moves the right side to the left

> A big downside with this action is that the frame image needs to be as long as the longest side of the image, which is why the input images have been resized beforehand.

![fWMvZ](https://user-images.githubusercontent.com/1164476/216736792-a004b1af-80f8-47a9-8ad6-9de5dee451ad.gif)

If you wanted just a black stroke around an image, that would be easy to do using actions and batch processing. Way more flexible than that Image Framing action. I found [this video tutorial](https://youtu.be/1esX_KTcblQ). I only quickly scrubbed through it, but it seemed pretty good.

As far as other scripts go, I'm not aware of any that would really help you with this, but if you do manage to preprocess the input files (resize+add a border/frame), which you could probably do with batch processing and actions, you could then feed those image files to this script with the smart object settings `resize: false, align: 'center bottom'`.

**Reply by joonaspaakko** [reply]
I was on a procrastination streak so I tried my own suggestion... and it does work with some caveats

Caveats:
1. The `Framing Action` is a bit cumbersome to use if you want to add multiple different frames
2. You can only do mite joints, so if you are emulating real frames that could be an issue
3. Generally speaking it's not easy to get realistic or good loking end results.
4. Related to the last point, the frame shadows need to be constructed using layer styles, which is not easy when you're trying to make realistic shadows... Unless you bake the shadows into the framed image files.
5. Unless your input images are very similar in size and aspect ratio, you can't do a mockup with multiple framed images per mockup. Or maybe you can if the frames aren't all like side by side.
6. You either need to use a mockup with an empty spot where you place the framed images... or you need to do some photo manipulation to like hide some objects.

[Dropbox link](https://www.dropbox.com/sh/7j70iu4628memkp/AACQJmirEzfI8GFCv6j-NjINa?dl=0)

Here's a vide owhere I go through the files so you can kinda see what needed to happen to get from the original photos to the output:

https://youtu.be/P3kqEyHFuQw

**Reply by spaceda1** [reply]
Hi Joonas
thanks you very much for those files.
Idea looks very good.
As far I was able to check it is two step processing piece of code.


After adding Framing Action it was throwing errors and was not able to finish adding frame to images.  It stops at "Place" command in the middle of action script.
Is it used as standalone action or called from main script?
looks it is stand alone to prepare framed images - but couldnt make it work.
See error I get. Opened base image and called an framing action... 
regs
z
![photoshop_XeQFKFfhah](https://user-images.githubusercontent.com/124141579/217948754-67c25a43-2f36-45cb-8c70-be189c48819d.jpg)

**Reply by spaceda1** [reply]
I was thinking also on combining frame action with your BMSOR script...or it is but I simply cant  make it work.

**Reply by spaceda1** [reply]
Got some time to make i working partially - I can successfully run action manually per single file but batch action breaks with error.
 
![rUNHgw97He](https://user-images.githubusercontent.com/124141579/217966069-0038c368-f14c-4922-a494-7518263a42c1.jpg)

**Reply by joonaspaakko** [reply]
> ~~My best guess is that you've changed a setting in the [`Layers` panel options](https://youtu.be/jkBs6KMLFws) that makes it so you don't get the added `Copy` text when duplicating layers (around 1:22 in the video). Another could be that your PS language is not English, but the dialogs indicate otherwise.~~ 

I wrote the original comment in the morning and I'm now realizing that isn't what the dialog is saying at all. For whatever reason it can't duplicate the current layer, which could be because the layer is locked or there aren't any layers selected at that moment. I would stop the action there and see if there are any obvious reasons why it couldn't duplicate the selected layer. And if I find the reason I'd work my way down the `action` one step at a time to see where it goes wrong. I can't think of a reason why it would do this. 

All of these types of issues are very common issues with `actions` shared with other people, since `actions` rely heavily on the exact environment they were recorded it. I might've said it before, but I'll say it again, `actions` can be pretty clunky.

> Like if you record an `action` where you close the document: if in your scenario PS doesn't ask you if you want to save the document because the document was already saved and someone runs the action but their document isn't saved at that point, it will stop the action right there. To avoid that, you'd need to record the action with that in mind. If while recording the action you get the "Want to save?" dialog and then close it, now it will not appear for the next person. 

----

Addressing some of the other things you said...

- You said it looks like a 2 step process: it depends on how you count the steps, but I'd say it's 3 steps minimum: 1. resize input images, 2. frame input images 3. Batch process input images through the mockup script
    - Resizing to the max size of the smart object before framing is important, unless you want every frame to be slightly different width (unless they are all same size). If the script's `resize` option is set to false, the frames would be the same width, but the entire framed image would have hard time fitting nicely inside the area reserved for your smart object.
- This is probably obvious to you at this point, but since you asked, the `action` is very much standalone and has nothing to do with the script.
- Combining the `action` with the script: I guess you could, but it's not that easy and at that point I think it might as well be written in JS instead to make things easier and more flexible.
    - I have considered adding certain points to the script where it would trigger an action if defined in the settings script, but it's not super common for batch processing operations. Also in most of those cases you can just divide it into a few batch process steps, like I have done here. I did say this already, but to really drive it in... If the script could trigger actions at certain points you'd still have to edit this specific action super heavily for it to work and more importantly to get any benefit from it (I think).

**Reply by spaceda1** [reply]
> I wrote the original comment in the morning and I'm now realizing that isn't what the dialog is saying at all. For whatever reason it can't duplicate the current layer, which could be because the layer is locked or there aren't any layers selected at that moment. I would stop the action there and see if there are any obvious reasons why it couldn't duplicate the selected layer. And if I find the reason I'd work my way down the action one step at a time to see where it goes wrong. I can't think of a reason why it would do this.

If I call manually the frame action it completes ok. But using it in batch stops with this error. this is the only difference.
so action itself looks ok but somehow it does not work with automated components.
I will try to make new layer via copy in my env and let you know.

yes you are right about steps . I was thinking about 2 steps as action script and your scrip which from my perspective are two separate things (cause I need to switch between it..I have >10k images.) Although I have some more automation tools and resizing is not a problem  - organizing all the pieces can be tricky and is quite vulnerable to even small difference between environments.  

thanks
will be back soon

**Reply by joonaspaakko** [reply]
I'm unable to replicate the issue... I don't think it'll do much good, but I can record a video of me using the action later today. Maybe you'll notice something in it. Otherwise, I think I'm out of ideas.

Actually, I did get another idea... Does it process any images or does it get stuck on the first one? I was thinking that perhaps you got lucky using the action on a single image... What if the difference isn't running the action as a batch process and what if there's an image or multiple images that don't allow layers or something? And I guess I could urge you to see what the status of the file looks like after it stops. That could help you work out what happened and maybe you could undo changes to see where it went wrong.

**Reply by joonaspaakko** [reply]
Here's [the video](https://youtu.be/tamaKFb8QNk). Not that I tried many different things, but I didn't get that dialog.

**Reply by spaceda1** [reply]
Hi
So I finally managed to get it working. There was strange problem in PS itself - had to reinstall and it worked.
thanks for this video cause it shows few differences between my config and what you have on your side.

there were also some differences in save file options as you had custom settings for save for web and it was set for image and on my side on windows it worked only with background save. strange. For whole process to be completed I also had to move close command to the main frame action thread cause it was not called at all (all images stayed open in PS).

I was looking at your   Frame Dark.psd and wonder what elements do I need to modify to add other width frames.
I will also include resize image inside frame action prior to framing saving manual operations before using actions.
thank you very much

**Reply by joonaspaakko** [reply]
Good. 

There's nothing too special when it comes to increasing the frame thickness. You'd likely want to use `Image > Canvas size...` and increase the height. It could be there's extra stuff overflowing the canvas, meaning that if you do want to use that specific frame, you might not have to do anything else aside from that. If there's no extra texture beyond the current canvas size, you'll have to just prepare the texture yourself. That said you could perhaps get away with increasing the height with `Image > Image size...` in this case.

The only special case about changing the frame height again comes from the fact that `actions` have static values. There's a step in the `action` that increases the `canvas size` to make room for the frame, so it doesn't cover too much of the image. That `action` step should increase the canvas by `canvas height - 1px`  (-1px just to make sure there won't be weird gapps). In my Dropbox test files I think the action is actually not increasing the canvas quite enough for how tall the frame is... I think. It's not a big deal if you don't care about showing 100% of the image. You could keep it as it is, but the thicker you make the frame, the more it's going to cover the framed image from the sides.

> The reason I had the close action separated was so that you can test framing a single image easily without it also saving and closing the document. I'm guessing you forgot to enable the very last step in the main `action` that triggers the other action. You can see it [in the video here](https://youtu.be/tamaKFb8QNk?t=58), where I'm changing the file paths, the very last step of the action is disabled (so it could be tested in a single image). Or maybe it had to do with you not wanting to save using save for web. Personally in that case I would've swapped that export with something else, but as long as it works for you, it doesn't really matter.

**Reply by spaceda1** [reply]
Definitely will check different frame sizes ! thank you.
there are thousands of images in a queue...but it looks your script will process them quite fast.

Did you test any perspective mockups?
regs

**Reply by joonaspaakko** [reply]
No I didn't test it in anything other than that one mockup. From a script point of view there's no difference no matter how the Smart Object is skewed, since the script just places the input file inside the Smart Object and resizes it to fit or fill inside it. But that said, if the mockup is poorly constructed or the placeholder graphic is a logo or something, it could end up looking weird if you place a photo there... Or it could look terrible with a perspective transformation, because picture frames don't tend to be paper thin, which is not a problem when viewed from the front.
